### PR TITLE
Start of control implementation

### PR DESCRIPTION
**Pull request by SouthEndMusic**
Solves issue #266.

### CODE REVIEW

File: docs/python/example_control.ipynb
Comment (visr) [inline docs/python/example_control.ipynb:1]: Don't forget to remove the output when you're done :)

File: core/src/bmi.jl [hunk 1]
```diff
@@ -74,10 +102,112 @@ function create_callbacks(
...
+        for target_node_id in outneighbors(control.graph, control_node_id)
+            set_control_params!(p, target_node_id, control_state_new)
+        end
+
+        control.control_state[control_node_id] = (control_state_new, integrator.t)
+    end
+end
+
+function set_control_params!(p::Parameters, node_id::Int, control_state::String)
+    if node_id in p.pump.node_id
```
Comment (evetion) [inline core/src/bmi.jl:202]: We probably need to add validation here.

**Comment by evetion** [comment]
We need to implement a Julia test for control to merge this, plus some documentation. The rest can be discussed in the refinement and added later.

File: core/src/bmi.jl [hunk 2]
```diff
@@ -54,9 +54,28 @@ function BMI.initialize(T::Type{Model}, config::Config)::Model
         config.solver.maxiters,
     )
 
+    set_initial_controlled_parameters(parameters.control, u0, integrator)
+
     return Model(integrator, config, saved_flow)
 end
 
+function set_initial_controlled_parameters(
```
Comment (evetion) [inline core/src/bmi.jl:62]: This seems a mutating function, might be good to add ! to the function name. If so, you might want to reorder the arguments too (mutating arguments first).

File: core/src/bmi.jl [hunk 3]
```diff
@@ -81,10 +100,122 @@ function create_callbacks(
...
+"""
+Get a value for a condition. Currently only supports getting levels from basins.
+"""
+function get_value(p::Parameters, node_id::Int, variable::String, u)
+    # TODO: Add support for getting flow values
+
+    if variable == "level"
+        basin = p.basin
+
+        # NOTE: Getting the level with get_level does NOT work
```
Comment (evetion) [inline core/src/bmi.jl:144]: Either remove or explain why (root finding needs interpolated values?)

File: core/src/create.jl [hunk 4]
```diff
@@ -85,7 +85,29 @@ end
 
 function Pump(db::DB, config::Config)::Pump
     static = load_structvector(db, config, PumpStaticV1)
-    return Pump(static.node_id, static.flow_rate)
+
+    control_mapping = Dict{Tuple{Int, String}, NamedTuple}()
+
+    if length(static.control_state) > 0 && !any(ismissing.(static.control_state))
+        # Starting flow_rates are first one found (can be updated by control initialisation)
```
Comment (evetion) [inline core/src/create.jl:92]: So the default state (normally missing) is not necessarily the one used, if the order of the static table is incorrect?

File: core/src/create.jl [hunk 5]
```diff
@@ -127,6 +149,42 @@ function Basin(db::DB, config::Config)::Basin
...
+    rows = execute(db, "select from_node_id, edge_type from Edge")
+    for (; from_node_id, edge_type) in rows
+        if edge_type == "control"
+            control_state[from_node_id] = ("undefined_state", 0.0)
+        end
+    end
+
+    logic = load_structvector(db, config, ControlLogicV1)
+
+    logic_mapping::Dict{Tuple{Int, String}, String} = Dict()
```
Comment (evetion) [inline core/src/create.jl:167]: I think you refactored this back, why not
`logic_mapping = Dict{Tuple{Int, String}, String}()`?

File: core/src/create.jl (hunk 4 above)
Comment (evetion) [reply core/src/create.jl:92]: I think the previous code making an `ismissing` mask (or a defaultmask, once we define an other default value), makes more sense, also in readability.

File: core/src/io.jl [hunk 6]
```diff
@@ -86,6 +86,10 @@ function load_structvector(
         nt = merge(nt, (; time = DateTime.(nt.time, dateformat"yyyy-mm-dd HH:MM:SS.s")))
     end
 
+    if hasfield(T, :control_state) && !hasfield(typeof(nt), :control_state)
```
Comment (evetion) [inline core/src/io.jl:89]: Is this required for backwards compatibility? Ideally we don't add hardcoded fields to this generic function. (The :time check above is already a bit ugly).

File: core/test/control.jl [hunk 7]
```diff
@@ -0,0 +1,18 @@
...
+
+    pump_control_mapping::Dict{Tuple{Int64, String}, NamedTuple} =
+        Dict((4, "off") => (flow_rate = 0.0,), (4, "on") => (flow_rate = 1.0e-5,))
+
+    logic_mapping::Dict{Tuple{Int64, String}, String} =
+        Dict((5, "TT") => "on", (5, "TF") => "off", (5, "FF") => "on", (5, "FT") => "off")
+
+    @test p.pump.control_mapping == pump_control_mapping
+    @test p.control.logic_mapping == logic_mapping
+end
```
Comment (evetion) [inline core/test/control.jl:18]: EOF

**Review by evetion — COMMENTED** [review]
Is the python code coverage something we can test with a small example, or do we wait for the larger use cases?

File: core/src/io.jl (hunk 6 above)
Comment (SouthEndMusic) [reply core/src/io.jl:89]: This is indeed needed otherwise `table = StructVector{T}(nt)` below throws an error.

**Pull request merged**